### PR TITLE
vm-monitor: Unset memory.high on start + refactor cgroup handling

### PR DESCRIPTION
**Pull request by sharnoff**
## Problem

Over the past couple days, we've had a couple VMs hit issues with postgres getting hit by memory.high throttling, even after #5303 was supposed to fix that. The tl;dr of those issues is that because vm-monitor startup sets the file cache size first, before interacting with the cgroup, cgroup throttling can mean we timeout connecting to the file cache and never reset the cgroup, even if memory has been upscaled since then.

See e.g.:

- https://neondb.slack.com/archives/C03F5SM1N02/p1695218132208249
- https://neondb.slack.com/archives/C03F5SM1N02/p1695314613696659

## Summary of changes

This PR adds an additional step into vm-monitor startup, where we first set the cgroup's memory.high value to 'max', removing the capacity for throttling. This preferable to just setting memory.high before the file cache, because it's theoretically possible that the new value of memory.high could still be less than the current memory usage, in which case postgres could continue to be throttled without sufficient memory events to relieve that.

Implementing this properly involved adding a method to our internal cgroup interface, and it seemed like there was duplicated functionality there, so this PR unifies that as well, making things a bit more consistent.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] ~~If it is a core feature, I have added thorough tests.~~
- [ ] ~~Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?~~
- [ ] ~~If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.~~

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by sharnoff** [comment]
Had to revert this branch so it's based on `HEAD^`; I think #5345 broke lots of tests, but at least this PR appears to be ok :sweat_smile:

**Review by koivunej — APPROVED** [review]
For me this sounds like it could work, at least from the PR desc and comments. Some smaller comments. On slack chatted about testing possibilities, would prefer to see these tests with a remotely resource usage controllable simple binary...

File: libs/vm_monitor/src/runner.rs [hunk 1]
```diff
@@ -106,6 +106,50 @@ impl Runner {
...
+
+            state.cgroup = Some(cgroup);
+        } else {
+            // *NOTE*: We need to forget the sender so that its drop impl does not get ran.
+            // This allows us to poll it in `Monitor::run` regardless of whether we
+            // are managing a cgroup or not. If we don't forget it, all receives will
+            // immediately return an error because the sender is droped and it will
+            // claim all select! statements, effectively turning `Monitor::run` into
+            // `loop { fail to receive }`.
+            mem::forget(requesting_send);
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply libs/vm_monitor/src/runner.rs:150]: Yeah, I also unfortunately didn't notice it the first time around, and haven't gotten around to fixing it yet. AFAICT this leaks ~64B every time, but... it's such a small number and only if the cgroup stuff is disabled (it never is), so we're probably fine, even though it's not very clean

File: libs/vm_monitor/src/runner.rs [hunk 2]
```diff
@@ -106,6 +106,50 @@ impl Runner {
...
 
+        // If we have both the cgroup and file cache integrations enabled, it's possible for
+        // temporary failures to result in cgroup throttling (from memory.high), that in turn makes
+        // it near-impossible to connect to the file cache (because it times out). Unfortunately,
+        // we *do* still want to determine the file cache size before setting the cgroup's
+        // memory.high, so it's not as simple as just swapping the order.
+        //
+        // Instead, the resolution here is that on startup, we temporarily unset memory.high, to
+        // allow any existing throttling to dissipate. It's a bit of a hacky solution, but helps
+        // with reliability.
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply libs/vm_monitor/src/runner.rs]: Already discussed on slack, but for future readers, updated the comment to make this a little more clear.

**Pull request merged**